src/Tests/TensorTest.cpp

```diff
@@ -0,0 +1,316 @@
+/*
+  ZynAddSubFX - a software synthesizer
+
+  TensorTest.cpp - Test for Tensor classes
+  Copyright (C) 2020-2022 Johannes Lorenz
+  Author: Johannes Lorenz
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+*/
+
+
+#include "../Params/WaveTable.h"
+#include <cstdlib>
+#include <cstdio>
+#include "test-suite.h"
+using namespace zyn;
+
+class TensorTest
+{
+        // replacements because TS_ASSERT_EQUALS requires a copy ctor
+        template <std::size_t N, class T>
+        void MY_TS_ASSERT_EQUALS(const Tensor<N, T>& t1, const Tensor<N, T>& t2)
+        {
+            TS_ASSERT(t1 == t2);
+        }
+        template <std::size_t N, class T>
+        void MY_TS_ASSERT_DIFFERS(const Tensor<N, T>& t1, const Tensor<N, T>& t2)
+        {
+            TS_ASSERT(t1 != t2);
+        }
+
+    public:
+
+        void setUp() {}
+        void tearDown() {}
+
+        void testPowerOf2(void)
+        {
+            TS_ASSERT_EQUAL_INT(AbstractRingbuffer::debugfunc_is_power_of_2(1), 1);
+            TS_ASSERT_EQUAL_INT(AbstractRingbuffer::debugfunc_is_power_of_2(2), 1);
+            TS_ASSERT_EQUAL_INT(AbstractRingbuffer::debugfunc_is_power_of_2(4), 1);
+            TS_ASSERT_EQUAL_INT(AbstractRingbuffer::debugfunc_is_power_of_2(0), 0);
+            TS_ASSERT_EQUAL_INT(AbstractRingbuffer::debugfunc_is_power_of_2(3), 0);
+        }
+
+        void testRb(void)
+        {
+            AbstractRingbuffer rb(4);
```

Comment:
More of a note for myself to look at on a second pass, but I can't recall if this particular ring buffer is just a circular buffer or if it's one with the threading semantics. IIRC adding the threading semantics might (implementation dependent) reduce the write space by 1.

Reply:
OK, I assume this is already solved with one of the comments below, so I resolve it here. Feel free to unresolve.